SNAPSHOT do not build for Bedrock

**Bug description**
Our project stopped building sometime in the last few days. We are using the snapshot version, as recommended in the docs. ~~We tried pinning it to the recently released M3 and had the same issues. We have different issues with the M2 release due to changes in the code contracts that are fixable but frustrating.~~ The M3 issue was something else

**Environment**
org.springframework.ai:spring-ai-bom:1.0.0-SNAPSHOT

 implementation 'org.springframework.ai:spring-ai-bedrock-ai-spring-boot-starter'

We are using  'org.springframework.boot' version '3.3.5'

**Steps to reproduce**
Try to build the spring AI SNAPSHOT with the org.springframework.ai:spring-ai-bedrock-ai-spring-boot-starter
`java.lang.IllegalStateException: Error processing condition on org.springframework.ai.autoconfigure.bedrock.converse.BedrockConverseProxyChatAutoConfiguration.bedrockProxyChatModel
	at org.springframework.boot.autoconfigure.condition.SpringBootCondition.matches(SpringBootCondition.java:60)
	at org.springframework.context.annotation.ConditionEvaluator.shouldSkip(ConditionEvaluator.java:108)
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsForBeanMethod(ConfigurationClassBeanDefinitionReader.java:183)
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsForConfigurationClass(ConfigurationClassBeanDefinitionReader.java:144)
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitions(ConfigurationClassBeanDefinitionReader.java:120)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:429)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:290)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:349)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:118)
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:789)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:607)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:754)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:456)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:335)
	at org.springframework.boot.test.context.SpringBootContextLoader.lambda$loadContext$3(SpringBootContextLoader.java:137)
	at org.springframework.util.function.ThrowingSupplier.get(ThrowingSupplier.java:58)
	at org.springframework.util.function.ThrowingSupplier.get(ThrowingSupplier.java:46)
	at org.springframework.boot.SpringApplication.withHook(SpringApplication.java:1463)
	at org.springframework.boot.test.context.SpringBootContextLoader$ContextLoaderHook.run(SpringBootContextLoader.java:553)
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:137)
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:108)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:225)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:152)
	... 17 more
Caused by: java.lang.IllegalStateException: @ConditionalOnMissingBean did not specify a bean using type, name or annotation and the attempt to deduce the bean's type failed
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition$Spec.validate(OnBeanCondition.java:515)
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition$Spec.<init>(OnBeanCondition.java:463)
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition.getMatchOutcome(OnBeanCondition.java:156)
	at org.springframework.boot.autoconfigure.condition.SpringBootCondition.matches(SpringBootCondition.java:47)
	... 39 more
Caused by: org.springframework.boot.autoconfigure.condition.OnBeanCondition$BeanTypeDeductionException: Failed to deduce bean type for org.springframework.ai.autoconfigure.bedrock.converse.BedrockConverseProxyChatAutoConfiguration.bedrockProxyChatModel
	at app//org.springframework.boot.autoconfigure.condition.OnBeanCondition$Spec.deducedBeanTypeForBeanMethod(OnBeanCondition.java:545)
	at app//org.springframework.boot.autoconfigure.condition.OnBeanCondition$Spec.deducedBeanType(OnBeanCondition.java:534)
	at app//org.springframework.boot.autoconfigure.condition.OnBeanCondition$Spec.<init>(OnBeanCondition.java:456)
	... 41 more
Caused by: java.lang.ClassNotFoundException: org.springframework.ai.bedrock.converse.BedrockProxyChatModel
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:467)
	at org.springframework.boot.autoconfigure.condition.FilteringSpringBootCondition.resolve(FilteringSpringBootCondition.java:108)
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition$Spec.getReturnType(OnBeanCondition.java:553)
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition$Spec.deducedBeanTypeForBeanMethod(OnBeanCondition.java:541)
	... 43 more`

**Expected behavior**
That it would build. If there was a breaking change that how to fix it would be communicated somewhere easy to find and easy to fix.

--- a/models/spring-ai-bedrock-converse/src/main/java/org/springframework/ai/bedrock/converse/BedrockProxyChatModel.java
+++ b/models/spring-ai-bedrock-converse/src/main/java/org/springframework/ai/bedrock/converse/BedrockProxyChatModel.java
@@ -36,7 +36,6 @@ import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 import reactor.core.publisher.Sinks;
 import reactor.core.publisher.Sinks.EmitFailureHandler;
-import reactor.core.publisher.Sinks.EmitResult;
 import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
 import software.amazon.awssdk.core.SdkBytes;
 import software.amazon.awssdk.core.document.Document;
@@ -524,6 +523,9 @@ public class BedrockProxyChatModel extends AbstractToolCallSupport implements Ch
 		});
 	}
 
+	public static final EmitFailureHandler DEFAULT_EMIT_FAILURE_HANDLER = EmitFailureHandler
+		.busyLooping(Duration.ofSeconds(10));
+
 	/**
 	 * Invoke the model and return the response stream.
 	 *
@@ -541,26 +543,19 @@ public class BedrockProxyChatModel extends AbstractToolCallSupport implements Ch
 		ConverseStreamResponseHandler.Visitor visitor = ConverseStreamResponseHandler.Visitor.builder()
 			.onDefault(output -> {
 				logger.debug("Received converse stream output:{}", output);
-				eventSink.tryEmitNext(output);
+				eventSink.emitNext(output, DEFAULT_EMIT_FAILURE_HANDLER);
 			})
 			.build();
 
 		ConverseStreamResponseHandler responseHandler = ConverseStreamResponseHandler.builder()
 			.onEventStream(stream -> stream.subscribe(e -> e.accept(visitor)))
 			.onComplete(() -> {
-				EmitResult emitResult = eventSink.tryEmitComplete();
-
-				while (!emitResult.isSuccess()) {
-					logger.info("Emitting complete:{}", emitResult);
-					emitResult = eventSink.tryEmitComplete();
-				}
-
-				eventSink.emitComplete(EmitFailureHandler.busyLooping(Duration.ofSeconds(3)));
+				eventSink.emitComplete(DEFAULT_EMIT_FAILURE_HANDLER);
 				logger.info("Completed streaming response.");
 			})
 			.onError(error -> {
 				logger.error("Error handling Bedrock converse stream response", error);
-				eventSink.tryEmitError(error);
+				eventSink.emitError(error, DEFAULT_EMIT_FAILURE_HANDLER);
 			})
 			.build();
 

--- a/models/spring-ai-bedrock/src/main/java/org/springframework/ai/bedrock/api/AbstractBedrockApi.java
+++ b/models/spring-ai-bedrock/src/main/java/org/springframework/ai/bedrock/api/AbstractBedrockApi.java
@@ -32,7 +32,6 @@ import org.slf4j.LoggerFactory;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Sinks;
 import reactor.core.publisher.Sinks.EmitFailureHandler;
-import reactor.core.publisher.Sinks.EmitResult;
 import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
 import software.amazon.awssdk.auth.credentials.ProfileCredentialsProvider;
 import software.amazon.awssdk.core.SdkBytes;
@@ -69,6 +68,10 @@ import org.springframework.util.Assert;
 public abstract class AbstractBedrockApi<I, O, SO> {
 
 	private static final Logger logger = LoggerFactory.getLogger(AbstractBedrockApi.class);
+
+	public static final EmitFailureHandler DEFAULT_EMIT_FAILURE_HANDLER = EmitFailureHandler
+	.busyLooping(Duration.ofSeconds(10));
+
 
 	private final String modelId;
 	private final ObjectMapper objectMapper;
@@ -264,7 +267,7 @@ public abstract class AbstractBedrockApi<I, O, SO> {
 			body = SdkBytes.fromUtf8String(this.objectMapper.writeValueAsString(request));
 		}
 		catch (JsonProcessingException e) {
-			eventSink.tryEmitError(e);
+			eventSink.emitError(e, DEFAULT_EMIT_FAILURE_HANDLER);
 			return eventSink.asFlux();
 		}
 
@@ -279,16 +282,16 @@ public abstract class AbstractBedrockApi<I, O, SO> {
 					try {
 						logger.debug("Received chunk: " + chunk.bytes().asString(StandardCharsets.UTF_8));
 						SO response = this.objectMapper.readValue(chunk.bytes().asByteArray(), clazz);
-						eventSink.tryEmitNext(response);
+						eventSink.emitNext(response, DEFAULT_EMIT_FAILURE_HANDLER);
 					}
 					catch (Exception e) {
 						logger.error("Failed to unmarshall", e);
-						eventSink.tryEmitError(e);
+						eventSink.emitError(e, DEFAULT_EMIT_FAILURE_HANDLER);
 					}
 				})
 				.onDefault(event -> {
 					logger.error("Unknown or unhandled event: " + event.toString());
-					eventSink.tryEmitError(new Throwable("Unknown or unhandled event: " + event.toString()));
+					eventSink.emitError(new Throwable("Unknown or unhandled event: " + event.toString()),DEFAULT_EMIT_FAILURE_HANDLER);
 				})
 				.build();
 
@@ -296,18 +299,12 @@ public abstract class AbstractBedrockApi<I, O, SO> {
 				.builder()
 				.onComplete(
 						() -> {
-							EmitResult emitResult = eventSink.tryEmitComplete();
-							while (!emitResult.isSuccess()) {
-								System.out.println("Emitting complete:" + emitResult);
-								emitResult = eventSink.tryEmitComplete();
-							}
-							eventSink.emitComplete(EmitFailureHandler.busyLooping(Duration.ofSeconds(3)));
-							// EmitResult emitResult = eventSink.tryEmitComplete();
-							logger.debug("\nCompleted streaming response.");
+							eventSink.emitComplete(DEFAULT_EMIT_FAILURE_HANDLER);
+							logger.info("Completed streaming response.");
 						})
 				.onError(error -> {
 					logger.error("\n\nError streaming response: " + error.getMessage());
-					eventSink.tryEmitError(error);
+					eventSink.emitError(error, DEFAULT_EMIT_FAILURE_HANDLER);
 				})
 				.onEventStream(stream -> stream.subscribe(
 						(ResponseStream e) -> e.accept(visitor)))

--- a/models/spring-ai-bedrock/src/test/java/org/springframework/ai/bedrock/cohere/BedrockCohereChatModelIT.java
+++ b/models/spring-ai-bedrock/src/test/java/org/springframework/ai/bedrock/cohere/BedrockCohereChatModelIT.java
@@ -23,6 +23,7 @@ import java.util.Map;
 import java.util.stream.Collectors;
 
 import com.fasterxml.jackson.databind.ObjectMapper;
+import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable;
 import reactor.core.publisher.Flux;
@@ -55,6 +56,7 @@ import static org.assertj.core.api.Assertions.assertThat;
 @SpringBootTest
 @EnabledIfEnvironmentVariable(named = "AWS_ACCESS_KEY_ID", matches = ".*")
 @EnabledIfEnvironmentVariable(named = "AWS_SECRET_ACCESS_KEY", matches = ".*")
+@Disabled("COHERE_COMMAND_V14 is not supported anymore")
 class BedrockCohereChatModelIT {
 
 	@Autowired

--- a/models/spring-ai-bedrock/src/test/java/org/springframework/ai/bedrock/cohere/api/CohereChatBedrockApiIT.java
+++ b/models/spring-ai-bedrock/src/test/java/org/springframework/ai/bedrock/cohere/api/CohereChatBedrockApiIT.java
@@ -19,6 +19,7 @@ package org.springframework.ai.bedrock.cohere.api;
 import java.time.Duration;
 import java.util.List;
 
+import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable;
 import reactor.core.publisher.Flux;
@@ -71,6 +72,7 @@ public class CohereChatBedrockApiIT {
 	}
 
 	@Test
+	@Disabled("Due to model version has reached the end of its life")
 	public void chatCompletion() {
 
 		var request = CohereChatRequest
@@ -95,6 +97,7 @@ public class CohereChatBedrockApiIT {
 		assertThat(response.generations().get(0).text()).isNotEmpty();
 	}
 
+	@Disabled("Due to model version has reached the end of its life")
 	@Test
 	public void chatCompletionStream() {
 

--- a/models/spring-ai-bedrock/src/test/java/org/springframework/ai/bedrock/jurassic2/BedrockAi21Jurassic2ChatModelIT.java
+++ b/models/spring-ai-bedrock/src/test/java/org/springframework/ai/bedrock/jurassic2/BedrockAi21Jurassic2ChatModelIT.java
@@ -158,8 +158,8 @@ class BedrockAi21Jurassic2ChatModelIT {
 			return new BedrockAi21Jurassic2ChatModel(jurassic2ChatBedrockApi,
 					BedrockAi21Jurassic2ChatOptions.builder()
 						.withTemperature(0.5)
-						.withMaxTokens(100)
-						.withTopP(0.9)
+						.withMaxTokens(500)
+						// .withTopP(0.9)
 						.build());
 		}
 

--- a/spring-ai-spring-boot-autoconfigure/src/main/java/org/springframework/ai/autoconfigure/bedrock/converse/BedrockConverseProxyChatAutoConfiguration.java
+++ b/spring-ai-spring-boot-autoconfigure/src/main/java/org/springframework/ai/autoconfigure/bedrock/converse/BedrockConverseProxyChatAutoConfiguration.java
@@ -51,7 +51,7 @@ import org.springframework.context.annotation.Import;
  */
 @AutoConfiguration
 @EnableConfigurationProperties({ BedrockConverseProxyChatProperties.class, BedrockAwsConnectionConfiguration.class })
-@ConditionalOnClass({ BedrockRuntimeClient.class, BedrockRuntimeAsyncClient.class })
+@ConditionalOnClass({ BedrockProxyChatModel.class, BedrockRuntimeClient.class, BedrockRuntimeAsyncClient.class })
 @ConditionalOnProperty(prefix = BedrockConverseProxyChatProperties.CONFIG_PREFIX, name = "enabled",
 		havingValue = "true", matchIfMissing = true)
 @Import(BedrockAwsConnectionConfiguration.class)

